src/http/request.rs

```diff
@@ -265,11 +270,24 @@ impl Request {
         Status::NGX_DONE
     }
 
+    /// how many subrequests are available to make in this request
+    /// will return NGX_HTTP_MAX_SUBREQUESTS for a parent request.
+    pub fn subrequests_available(&self) -> u32 {
+        // 1 is subtracted because this function was caught returning 1 extra
+        // The return value should be (50, 0), with the parent request returning
+        // NGX_HTTP_MAX_SUBREQUESTS.
+        // See http://nginx.org/en/docs/dev/development_guide.html#http_subrequests
+        // for more information
+        self.0.subrequests() - 1
+    }
+
     /// Send a subrequest
     pub fn subrequest(
```

Comment:
Question to self: Why not support args? If needed later it's an API change. If needed, will add.

Reply:
@ivanitskiy I notice that internal_redirect, the wrapper for ngx_http_internal_redirect doesn't support `args` either.

i see modules do use this feature, often pulled from the Request itself, but not every case.

can we create a `internal_redirect_with_args` which `internal_redirect` calls - that way `internal_redirect` signature remains stable and we open up args to module writers?

Reply:
good question and a good catch. 

I wrote rust fn `internal_redirect` with an intention to combine "named_location" and "internal redirect" functionally. the use-case for a module i had was to implement something similar to `http_auth_request_module` of nginx config and allow using named and regular locations:
```
http {
    #...
    server {
    #...
        location /private/ {
            auth_request     /auth;
           # auth_request     @named_loc_instead;
        }

        location = /auth {
            internal;
            #<....>
        }
    }
}
```

in my understanding `args` are URI arguments (the one that go after`?`, and look like `foo=bar`. as I understand those are part of URI as well. so I guess we can extract them from a) `location` b) expose in the fn as Optin<&str>

`pub fn internal_redirect(&self, location: &str, args: Option<&str>) -> Status {`

Reply:
another option is to have two rust functions that are "safe" and wraps a C call:

```rust
pub fn named_location(&self, location: &str) -> Status {
        ngx_http_named_location((self as *const Request as *mut Request).cast(), uri_ptr);
  }      
```
and 
```rust
pub fn _internal_redirect(&self, location: &str, args: ???) -> Status {
                ngx_http_internal_redirect(
                    (self as *const Request as *mut Request).cast(),
                    uri_ptr,
                    ??? // std::ptr::null_mut(),
                )
}
```

Reply:
That was my conclusion too, `args` is the query and fragments. Args can be pulled from the URI, but my assumption for the argument was to provide flexibility and mutatable args - I do see a case or two where they are passed in separate from what's stored on the request.

Adding `args` to internal_redirect is a clean approach going forward, but it will break existing users. Are we ok with that seeing it's early? I don't want to make it a habit but, on the other hand, it's still a young project.

Reply:
This is part of the "stabilization" process being pre-v1 versions. 

I can see both ways as a go or a combination.
I can offer this as it offers all variants and a backward compatible change (and maybe need a bit fancier doc to explain the differences/ not sure about fn names too):

```rust
    /// Perform internal redirect without args or named location
    /// see `ngx_internal_redirect` and `ngx_http_named_location`
    pub fn internal_redirect(&self, location: &str) -> Status {
        if location.starts_with('@') {
            self.ngx_http_named_location(location)
        } else {
            self.ngx_internal_redirect(location, "")
        }
    }

    /// Invoke ngx_internal_redirect to perform an internal redirect to a location
    pub fn ngx_internal_redirect(&self, location: &str, args: &str) -> Status {
        assert!(!location.is_empty(), "uri location is empty");
        let uri_ptr = &mut ngx_str_t::from_str(self.0.pool, location) as *mut _;
        let args_ptr = if args.len() > 0 {
            &mut ngx_str_t::from_str(self.0.pool, location) as *mut _
        } else {
            std::ptr::null_mut()
        };

        unsafe {
            Status(ngx_http_internal_redirect(
                (self as *const Request as *mut Request).cast(),
                uri_ptr,
                std::ptr::null_mut(),
            ))
        }
    }

    /// Invoke ngx_http_named_location
    pub fn ngx_http_named_location(&self, location: &str) -> Status {
        assert!(!location.is_empty(), "uri location is empty");
        assert!(location.starts_with('@'), "named location must start with @");
        let uri_ptr = &mut ngx_str_t::from_str(self.0.pool, location) as *mut _;

        unsafe {
            Status(ngx_http_named_location(
                (self as *const Request as *mut Request).cast(),
                uri_ptr,
            ))
        }
    }
```

Reply:
note to me: maybe there is a better way to obtain a raw pointer to a str. currently it copies data and allocates a new ngx_str by `let uri_ptr = &mut ngx_str_t::from_str(self.0.pool, location) as *mut _`

Reply:
good feedback, i'll make some edits today and get an update out.